File: src/main/kotlin/ninja/blacknet/transaction/CreateMultisig.kt
/*
 * Copyright (c) 2018-2019 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.transaction

import kotlinx.serialization.Serializable
import ninja.blacknet.contract.MultiSignatureLockContractId
import ninja.blacknet.core.*
import ninja.blacknet.crypto.*
import ninja.blacknet.crypto.Blake2b.buildHash
import ninja.blacknet.serialization.BinaryDecoder
import ninja.blacknet.serialization.BinaryEncoder
import ninja.blacknet.serialization.Json
import ninja.blacknet.serialization.LongSerializer
import ninja.blacknet.util.sumByLong

/**
 * 創建合約
 */
@Serializable
class CreateMultisig(
        val n: Byte,
        val deposits: ArrayList<DepositElement>,
        val signatures: ArrayList<SignatureElement>
) : TxData {
    @Serializable
    class DepositElement(
            val from: PublicKey,
            @Serializable(with = LongSerializer::class)
            val amount: Long
    ) {
        operator fun component1() = from
        operator fun component2() = amount
    }

    @Serializable
    class SignatureElement(
            val index: Byte,
            @Serializable(with = SignatureSerializer::class)
            val signature: ByteArray
    ) {
        operator fun component1() = index
        operator fun component2() = signature
    }

    fun id(hash: Hash, dataIndex: Int) =
        MultiSignatureLockContractId(
            buildHash { encodeHash(hash); encodeInt(dataIndex); }
        )

    fun sign(from: PublicKey, seq: Int, dataIndex: Int, privateKey: PrivateKey): Boolean {
        val publicKey = privateKey.toPublicKey()
        val index = deposits.indexOfFirst { it.from == publicKey }
        if (index == -1) return false
        val signature = Ed25519.sign(hash(from, seq, dataIndex), privateKey)
        signatures.add(SignatureElement(index.toByte(), signature))
        return true
    }

    private fun hash(from: PublicKey, seq: Int, dataIndex: Int): Hash {
        val copy = CreateMultisig(n, deposits, ArrayList())
        val bytes = BinaryEncoder.toBytes(serializer(), copy)
        return buildHash {
            encodePublicKey(from)
            encodeInt(seq)
            encodeInt(dataIndex)
            encodeByteArray(bytes)
        }
    }

    override fun processImpl(tx: Transaction, hash: Hash, dataIndex: Int, ledger: Ledger): Status {
        if (n < 0 || n > deposits.size) {
            return Invalid("Invalid n")
        }
        if (deposits.size > 20) {
            return Invalid("Too many deposits")
        }
        if (signatures.size > deposits.size) {
            return Invalid("Too many signatures")
        }
        val total = try {
            deposits.sumByLong { it.amount }
        } catch (e: ArithmeticException) {
            return Invalid("Invalid total amount: ${e.message}")
        }
        if (total <= 0) {
            return Invalid("Invalid total amount")
        }

        val multisigHash = hash(tx.from, tx.seq, dataIndex)
        for (index in 0 until deposits.size) {
            val (publicKey, amount) = deposits[index]
            if (amount != 0L) {
                val signature = signatures.find { it.index == index.toByte() }?.signature
                if (signature == null) {
                    return Invalid("Unsigned deposit $index")
                }
                if (!Ed25519.verify(signature, multisigHash, publicKey)) {
                    return Invalid("Invalid signature $index")
                }
                val depositAccount = ledger.get(publicKey)
                if (depositAccount == null) {
                    return Invalid("Account not found $index")
                }
                val status = depositAccount.credit(amount)
                if (status != Accepted) {
                    return notAccepted("CreateMultisig $index", status)
                }
                ledger.set(publicKey, depositAccount)
            }
        }

        val id = id(hash, dataIndex)
        val multisig = Multisig(n, deposits.map { (from, amount) -> Pair(from, amount) })
        ledger.addMultisig(id, multisig)
        return Accepted
    }

    fun involves(publicKey: PublicKey) = deposits.find { it.from == publicKey } != null
}

File: kernel/src/main/rust/transaction/createmultisig.rs
/*
 * Copyright (c) 2018-2025 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use crate::amount::Amount;
use crate::blake2b::{Blake2b256, Hash};
use crate::ed25519::{PublicKey, Signature, verify};
use crate::error::{Error, Result};
use crate::multisig::{Deposit, Multisig};
use crate::transaction::{CoinTx, Transaction, TxData};
use alloc::borrow::ToOwned;
use alloc::boxed::Box;
use blacknet_serialization::format::to_bytes;
use digest::Digest;
use serde::{Deserialize, Serialize};

pub type MultiSignatureLockContractId = [u8; 32];

fn id(hash: Hash, data_index: u32) -> MultiSignatureLockContractId {
    let mut hasher = Blake2b256::new();
    hasher.update(hash);
    hasher.update(data_index.to_be_bytes());
    hasher.finalize().into()
}

#[derive(Clone, Copy, Deserialize, Serialize)]
pub struct Dep {
    from: PublicKey,
    amount: Amount,
}

impl Dep {
    pub const fn from(self) -> PublicKey {
        self.from
    }

    pub const fn amount(self) -> Amount {
        self.amount
    }
}

#[derive(Clone, Copy, Deserialize, Serialize)]
pub struct Sig {
    index: u8,
    signature: Signature,
}

impl Sig {
    pub const fn index(self) -> u8 {
        self.index
    }

    pub const fn signature(self) -> Signature {
        self.signature
    }
}

#[derive(Deserialize, Serialize)]
pub struct CreateMultisig {
    n: u8,
    deposits: Box<[Dep]>,
    signatures: Box<[Sig]>,
}

impl CreateMultisig {
    pub const fn new(n: u8, deposits: Box<[Dep]>, signatures: Box<[Sig]>) -> Self {
        Self {
            n,
            deposits,
            signatures,
        }
    }

    pub const fn n(&self) -> u8 {
        self.n
    }

    pub const fn deposits(&self) -> &[Dep] {
        &self.deposits
    }

    pub const fn signatures(&self) -> &[Sig] {
        &self.signatures
    }

    fn hash(&self, from: PublicKey, seq: u32, data_index: u32) -> Result<Hash> {
        let copy = Self {
            n: self.n,
            deposits: self.deposits.clone(),
            signatures: Default::default(),
        };
        let bytes = to_bytes::<CreateMultisig>(&copy)?;
        let mut hasher = Blake2b256::new();
        hasher.update(from);
        hasher.update(seq.to_be_bytes());
        hasher.update(data_index.to_be_bytes());
        hasher.update(bytes);
        Ok(hasher.finalize().into())
    }
}

impl TxData for CreateMultisig {
    fn process_impl(
        &self,
        tx: Transaction,
        hash: Hash,
        data_index: u32,
        coin_tx: &mut (impl CoinTx + ?Sized),
    ) -> Result<()> {
        if self.n as usize > self.deposits.len() {
            return Err(Error::Invalid("Invalid n".to_owned()));
        }
        if self.deposits.len() > 20 {
            return Err(Error::Invalid("Too many deposits".to_owned()));
        }
        if self.signatures.len() > self.deposits.len() {
            return Err(Error::Invalid("Too many signatures".to_owned()));
        }
        match Amount::checked_sum(self.deposits.iter().copied().map(Dep::amount)) {
            Some(sum) => {
                if sum == Amount::ZERO {
                    return Err(Error::Invalid("Invalid total amount".to_owned()));
                }
            }
            None => return Err(Error::Invalid("Invalid total amount".to_owned())),
        };

        let multisig_hash = self.hash(tx.from(), tx.seq(), data_index)?;
        for (index, deposit) in self.deposits.iter().enumerate() {
            if deposit.amount() != Amount::ZERO {
                let sig = self
                    .signatures
                    .iter()
                    .find(|sig| sig.index == index as u8)
                    .ok_or(Error::Invalid("Unsigned deposit".to_owned()))?;
                verify(sig.signature(), multisig_hash, deposit.from())?;
                let mut deposit_account = coin_tx.get_account(deposit.from())?;
                deposit_account.credit(deposit.amount())?;
                coin_tx.set_account(deposit.from(), deposit_account);
            }
        }

        let id = id(hash, data_index);
        let multisig = Multisig::new(
            self.n,
            self.deposits
                .iter()
                .map(|dep| Deposit::new(dep.from(), dep.amount()))
                .collect(),
        );
        coin_tx.add_multisig(id, multisig);
        Ok(())
    }
}
